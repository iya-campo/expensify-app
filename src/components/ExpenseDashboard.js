import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpenseSummary';

const ExpenseDashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
