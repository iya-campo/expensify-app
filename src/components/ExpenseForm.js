import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? props.expense.amount : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide a description and amount.',
      }));
      if (!this.state.description && this.state.amount) {
        this.setState(() => ({
          error: 'Please provide a description.',
        }));
      } else if (this.state.description && !this.state.amount) {
        this.setState(() => ({
          error: 'Please provide an amount.',
        }));
      }
    } else {
      this.props.onSubmit({
        description: this.state.description,
        // amount: parseFloat(this.state.amount, 10) * 100, doesn't work?
        amount: Math.round(this.state.amount * 100) / 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(),
      });
      this.setState(() => ({
        error: '',
      }));
    }
  };
  render() {
    return (
      <form className='form' action='' onSubmit={this.onSubmit}>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          className='text-input'
          type='text'
          placeholder='Description'
          value={this.state.description}
          onChange={this.onDescriptionChange}
          autoFocus
        />
        <input className='text-input' type='text' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className='textarea'
          type='text'
          placeholder='Add a note for your expense... (Optional)'
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className='btn'>Save Expense</button>
        </div>
      </form>
    );
  }
}
