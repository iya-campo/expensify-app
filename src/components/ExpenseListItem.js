import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <div className='card'>
        <h3>
          {description}
          <Link className='ml-3' to={`/edit/${id}`}>
            Edit
          </Link>
        </h3>
        <p>Amount: {amount}</p>
        <p>{moment(createdAt).format('MMM Do, YYYY')}</p>
      </div>
    </div>
  );
};

export default ExpenseListItem;
