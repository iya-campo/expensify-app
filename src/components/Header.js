import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <ul className='list-group'>
      <li className='list-group-item'>
        <NavLink to='/' activeClassName='is-active' exact>
          Dashboard
        </NavLink>
      </li>
      <li className='list-group-item'>
        <NavLink to='/create' activeClassName='is-active'>
          Add Expense
        </NavLink>
      </li>
    </ul>
  </header>
);

export default Header;
