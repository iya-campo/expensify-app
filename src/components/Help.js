import React from 'react';

const Help = () => (
  <div>
    <p>This is from my help component.</p>
  </div>
);

export default Help;
