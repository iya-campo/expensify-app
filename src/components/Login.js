import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const Login = ({ startLogin }) => {
  return (
    <div>
      Login Page
      <button onClick={startLogin}>Login</button>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(Login);
