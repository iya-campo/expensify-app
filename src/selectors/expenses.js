import moment from 'moment';

// Get Visible Expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      // 1: Asc -1: Desc
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
