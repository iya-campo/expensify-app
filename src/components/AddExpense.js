import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpense extends React.Component {
  onSubmit = (expense) => {
    // Abstracting dispatch for easier testing
    // props.dispatch(addExpense(expense));
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpense);
