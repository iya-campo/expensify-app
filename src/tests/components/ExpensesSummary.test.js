import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpenseSummary';
import selectExpensesTotal from '../../selectors/expenses-total';

test('should render ExpensesSummary with one expense correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={[expenses[0]].length} expensesTotal={selectExpensesTotal([expenses[0]])} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expensesTotal={selectExpensesTotal(expenses)} />);
  expect(wrapper).toMatchSnapshot();
});
