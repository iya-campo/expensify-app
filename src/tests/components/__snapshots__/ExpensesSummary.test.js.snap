// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple expenses correctly 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $114,195.00
  </h1>
</div>
`;

exports[`should render ExpensesSummary with one expense correctly 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $195.00
  </h1>
</div>
`;
